lib/src/game/camera.dart

```diff
@@ -0,0 +1,289 @@
+import 'dart:math' as math;
+import 'dart:ui' show Rect;
+
+import '../../components.dart';
+import '../../game.dart';
+
+/// Utility method to smoothly transition vectors.
+void _moveToTarget(
+  Vector2 current,
+  Vector2 target,
+  double ds,
+) {
+  if (current != target) {
+    final diff = target - current;
+    if (diff.length < ds) {
+      current.setFrom(target);
+    } else {
+      diff.scaleTo(ds);
+      current.setFrom(current + diff);
+    }
+  }
+}
+
+/// A camera translates your game coordinate system; this is useful when your
+/// world is not 1:1 with your screen size.
+///
+/// A camera always has a current [position], however you cannot set it
+/// directly. You must use some methods to ensure that the camera moves smoothly
+/// as the game runs. Smoothly here means that sudden snaps should be avoided,
+/// as they feel jarring to the player.
+///
+/// There are three major factors that determine the camera position:
+///
+/// * Follow
+/// If you want, you can call [followObject] at the beginning of your
+/// stage/world/level, and provided a [PositionComponent].
+/// The camera will follow this component making sure its position is fixed
+/// on the screen.
+/// You can set the relative position of the screen you want the follow
+/// object to stay in (normally the center), and you can even change that
+/// and get a smooth transition.
+///
+/// * Move
+/// You can alternatively move the camera to a specific world coordinate.
+/// This will set the top left of the camera and will ignore any existing follow
+/// rules and move the camera smoothly until it reaches the desired destination.
+///
+/// * Shake
+/// Regardless of the the previous rules, you can additionally add a shake
+/// effect for a brief period of time on top of the current coordinate.
+/// The shake adds a random immediate delta to each tick to simulate the shake
+/// effect.
+///
+/// Note: in the context of the BaseGame, the camera effectively translates
+/// the position where components are rendered with relation to the Viewport.
+/// Components marked as `isHud = true` are always rendered in screen
+/// coordinates, bypassing the camera altogether.
+///
+/// Note: right now this only applies to rendering. No transformation is
+/// done on event handling (for gestures). You have to transform it yourself
+/// using [screenToWorld].
+class Camera {
+  static const DEFAULT_CAMERA_SPEED = 50.0; // in pixels/s
+  static const DEFAULT_SHAKE_INTENSITY = 75.0; // in pixels
+  static const DEFAULT_SHAKE_DURATION = 0.3; // in seconds
```

Comment:
Consts in dart shouldn't be java style
https://dart.dev/guides/language/effective-dart/style#prefer-using-lowercamelcase-for-constant-names

Reply:
WHAAAAT? I did **not** know that at all hahahaha

Shall we add this linter rule then? and amend the styleguide?